### PR TITLE
MULE-9550: Create domain descriptor for loader overrides and redeploy…

### PR DESCRIPTION
**Pull request by afelisatti**
…ment control

### CODE REVIEW

File: modules/launcher/src/main/java/org/mule/module/launcher/DefaultAppBloodhound.java [hunk 1]
```diff
@@ -67,35 +67,16 @@ public DefaultAppBloodhound()
...
-        Collection<File> deployFiles = FileUtils.listFiles(appDir, new WildcardFileFilter("mule-deploy.*"), null);
-        if (deployFiles.size() > 1)
-        {
-            // TODO need some kind of an InvalidAppFormatException
-            throw new MuleRuntimeException(
-                    MessageFactory.createStaticMessage(
-                            String.format("More than one mule-deploy descriptors found in application '%s'", appName)));
-        }
-
+        final File descriptorFile = ArtifactFactoryUtils.getDeploymentFile(appsDir, appName);
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/DefaultAppBloodhound.java]: Use static imports

Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/DefaultAppBloodhound.java]: You can use MuleFolderUtils#getAppFolder instead of passing appsDir and and name to getDeploymentFile method

File: modules/launcher/src/main/java/org/mule/module/launcher/DefaultAppBloodhound.java [hunk 2]
```diff
@@ -109,9 +90,10 @@ public ApplicationDescriptor fetch(String appName) throws IOException
 
             desc = descriptorParser.parse(descriptorFile, appName);
             // app name is external to the deployment descriptor
-            desc.setAppName(appName);
+            desc.setName(appName);
         }
 
+        File appDir = new File(appsDir, appName);
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/DefaultAppBloodhound.java]: Same as before

File: modules/launcher/src/main/java/org/mule/module/launcher/DeploymentDirectoryWatcher.java [hunk 3]
```diff
@@ -522,20 +522,26 @@ private void deleteAnchorsFromDirectory(final File directory)
 
     private void redeployModifiedDomains()
     {
-        redeployModifiedArtifacts(domains, domainTimestampListener, domainArchiveDeployer);
+        Collection redeployableDomains = getArtifactsToRedeploy(domains);
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/DeploymentDirectoryWatcher.java]: This changes look unrelated to the new feature

File: modules/launcher/src/main/java/org/mule/module/launcher/MuleSharedDomainClassLoader.java [hunk 4]
```diff
@@ -30,10 +31,15 @@
     private File domainDir;
     private File domainLibraryFolder;
 
-    @SuppressWarnings("unchecked")
     public MuleSharedDomainClassLoader(String domain, ClassLoader parent)
     {
-        super(new URL[0], parent);
+        this(domain, parent, null);
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/MuleSharedDomainClassLoader.java]: Pass an empty set instead of null

File: modules/launcher/src/main/java/org/mule/module/launcher/artifact/Artifact.java [hunk 5]
```diff
@@ -49,6 +50,11 @@
     String getArtifactName();
 
     /**
+     * @return the artifact descriptor
+     */
+    ArtifactDescriptor getDescriptor();
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/artifact/Artifact.java]: Cannot be the artifact be generic on the ArtifactDescriptor type?

File: modules/launcher/src/main/java/org/mule/module/launcher/artifact/ArtifactFactoryUtils.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import org.mule.api.MuleRuntimeException;
+import org.mule.config.i18n.MessageFactory;
+import org.mule.util.FileUtils;
+
+import java.io.File;
+import java.util.Collection;
+
+import org.apache.commons.io.filefilter.WildcardFileFilter;
+
+public class ArtifactFactoryUtils
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/artifact/ArtifactFactoryUtils.java]: Add javadoc

File: modules/launcher/src/main/java/org/mule/module/launcher/artifact/ArtifactFactoryUtils.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+{
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/artifact/ArtifactFactoryUtils.java]: Add private constructor

File: modules/launcher/src/main/java/org/mule/module/launcher/artifact/ArtifactFactoryUtils.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    /**
+     * @param lookupDirectory the directory where the artifact should be present
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/artifact/ArtifactFactoryUtils.java]: Add method description

File: modules/launcher/src/main/java/org/mule/module/launcher/artifact/ArtifactFactoryUtils.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        File artifactDir = new File(lookupDirectory, artifactName);
+        if (!artifactDir.exists())
+        {
+            throw new MuleRuntimeException(
+                    MessageFactory.createStaticMessage(
+                            String.format("Artifact directory does not exist: '%s'", artifactDir)));
+        }
+        // list mule-deploy.* files
+        @SuppressWarnings("unchecked")
+        Collection<File> deployFiles = FileUtils.listFiles(artifactDir, new WildcardFileFilter("mule-deploy.*"), null);
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/artifact/ArtifactFactoryUtils.java]: Mule supports, in theory, multiple configuration files formats. But reality is that we only use properties and I don't think that anybody is using a different type. I'm not sure if the support for apps could be removed, but for sure it is not needed on domains

File: modules/launcher/src/main/java/org/mule/module/launcher/artifact/ArtifactFactoryUtils.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        if (deployFiles.size() > 1)
+        {
+            // TODO need some kind of an InvalidAppFormatException
+            throw new MuleRuntimeException(
+                    MessageFactory.createStaticMessage(
+                            String.format("More than one mule-deploy descriptors found in artifact '%s'", artifactName)));
+        }
+
+        if(deployFiles.isEmpty())
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/artifact/ArtifactFactoryUtils.java]: Fix code style

File: modules/launcher/src/main/java/org/mule/module/launcher/descriptor/ArtifactDescriptor.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.module.launcher.descriptor;
+
+import java.util.HashSet;
+import java.util.Set;
+
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/descriptor/ArtifactDescriptor.java]: Add javadoc

File: modules/launcher/src/main/java/org/mule/module/launcher/descriptor/DomainDescriptor.java [hunk 12]
```diff
@@ -0,0 +1,11 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.module.launcher.descriptor;
+
+public class DomainDescriptor extends ArtifactDescriptor
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/descriptor/DomainDescriptor.java]: Is this class really needed?

File: modules/launcher/src/main/java/org/mule/module/launcher/descriptor/EmptyDomainDescriptor.java [hunk 13]
```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.module.launcher.descriptor;
+
+public class EmptyDomainDescriptor extends DomainDescriptor
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/descriptor/EmptyDomainDescriptor.java]: This class does not add too much. Add javadoc if you need to keep it

File: modules/launcher/src/main/java/org/mule/module/launcher/domain/DefaultDomainFactory.java [hunk 14]
```diff
@@ -49,13 +54,39 @@ public Domain createArtifact(String artifactName) throws IOException
...
     }
 
+    private DomainDescriptor findDomain(String domainName) throws IOException
+    {
+        if (DEFAULT_DOMAIN_NAME.equals(domainName))
+        {
+            return new EmptyDomainDescriptor(DEFAULT_DOMAIN_NAME);
+        }
+
+        final File appsDir = MuleContainerBootstrapUtils.getMuleDomainsDir();
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/domain/DefaultDomainFactory.java]: Use static imports and use MulefolderUtils

File: modules/launcher/src/main/java/org/mule/module/launcher/domain/DefaultMuleDomain.java [hunk 15]
```diff
@@ -124,7 +125,7 @@ public ConfigurationBuilder createApplicationConfigurationBuilder(Application ap
     protected String determineConfigBuilderClassNameForApplication(Application defaultMuleApplication)
     {
         // Provide a shortcut for Spring: "-builder spring"
-        final String builderFromDesc = defaultMuleApplication.getDescriptor().getConfigurationBuilder();
+        final String builderFromDesc = ((ApplicationDescriptor) defaultMuleApplication.getDescriptor()).getConfigurationBuilder();
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/domain/DefaultMuleDomain.java]: Ugly casting

File: modules/launcher/src/main/java/org/mule/module/launcher/domain/DomainDescriptorParser.java [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+import org.apache.commons.lang.BooleanUtils;
+
+/**
+ * Descriptor parser that only handles loader overrides and redeployment properties, the only ones currently supported by domains.
+ */
+public class DomainDescriptorParser
+{
+
+    private static final String PROPERTY_LOADER_OVERRIDE = "loader.override";
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/domain/DomainDescriptorParser.java]: Are not these properties already defined in another class?

File: modules/launcher/src/main/java/org/mule/module/launcher/domain/DomainDescriptorParser.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Properties;
+import java.util.Set;
+
+import org.apache.commons.lang.BooleanUtils;
+
+/**
+ * Descriptor parser that only handles loader overrides and redeployment properties, the only ones currently supported by domains.
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/domain/DomainDescriptorParser.java]: I would not list which properties are supported on the javadoc.

File: modules/launcher/src/main/java/org/mule/module/launcher/domain/DomainDescriptorParser.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+ */
+public class DomainDescriptorParser
+{
+
+    private static final String PROPERTY_LOADER_OVERRIDE = "loader.override";
+    private static final String PROPERTY_REDEPLOYMENT_ENABLED = "redeployment.enabled";
+
+    public DomainDescriptor parse(File descriptor) throws IOException
+    {
+        final Properties properties = PropertiesUtils.loadProperties(new FileInputStream(descriptor));
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/domain/DomainDescriptorParser.java]: Use static imports

File: modules/launcher/src/test/java/org/mule/module/launcher/DeploymentServiceTestCase.java [hunk 19]
```diff
@@ -2232,6 +2233,28 @@ public void redeploysDomainAndItsApplications() throws Exception
     }
 
     @Test
+    public void doesNotRedeployDomainWithRedeploymentDisabled() throws Exception
```
Comment (pablokraan) [inline modules/launcher/src/test/java/org/mule/module/launcher/DeploymentServiceTestCase.java]: This seems to be unrelated to the loader override change

File: modules/launcher/src/test/java/org/mule/module/launcher/builder/DomainFileBuilder.java [hunk 20]
```diff
@@ -93,6 +98,22 @@ public DomainFileBuilder containing(ApplicationFileBuilder appFileBuilder)
...
+     * Adds a property into the domain deployment properties file.
+     *
+     * @param propertyName name fo the property to add. Non empty
+     * @param propertyValue value of the property to add. Non null.
+     * @return the same builder instance
+     */
+    public DomainFileBuilder deployedWith(String propertyName, String propertyValue)
+    {
+        checkImmutable();
+        Preconditions.checkArgument(!StringUtils.isEmpty(propertyName), "Property name cannot be empty");
```
Comment (pablokraan) [inline modules/launcher/src/test/java/org/mule/module/launcher/builder/DomainFileBuilder.java]: Use static import and use the mule version of checkArgument

File: modules/launcher/src/main/java/org/mule/module/launcher/DeploymentDirectoryWatcher.java (hunk 3 above)
Comment (afelisatti) [inline modules/launcher/src/main/java/org/mule/module/launcher/DeploymentDirectoryWatcher.java]: This is were the "redeployment.enabled" property takes effect for domains.

File: modules/launcher/src/test/java/org/mule/module/launcher/domain/AbstractDomainTestCase.java [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.rules.TemporaryFolder;
+
+public class AbstractDomainTestCase extends AbstractMuleTestCase
+{
+    @ClassRule
+    public static TemporaryFolder temporaryFolder = new TemporaryFolder();
+    @Rule
+    public final SystemProperty muleHomeSystemProperty = new SystemProperty(MuleProperties.MULE_HOME_DIRECTORY_PROPERTY, temporaryFolder.getRoot().getCanonicalPath());
```
Comment (pablokraan) [inline modules/launcher/src/test/java/org/mule/module/launcher/domain/AbstractDomainTestCase.java]: Why temporaryFolder is static and muleHomeSystemProperty not, if the second will be always the same?

File: modules/launcher/src/test/java/org/mule/module/launcher/domain/DefaultDomainFactoryTestCase.java [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+import org.hamcrest.Matcher;
+import org.junit.Test;
+
+public class DefaultDomainFactoryTestCase extends AbstractDomainTestCase
+{
+    private DomainFactory domainFactory;
+
+    public DefaultDomainFactoryTestCase() throws IOException
+    {
+        this.domainFactory = new DefaultDomainFactory(new MuleDomainClassLoaderRepository());
```
Comment (pablokraan) [inline modules/launcher/src/test/java/org/mule/module/launcher/domain/DefaultDomainFactoryTestCase.java]: move initialization to field  declaration

File: modules/launcher/src/main/java/org/mule/module/launcher/artifact/ArtifactFactoryUtils.java (hunk 9 above)
Comment (afelisatti) [inline modules/launcher/src/main/java/org/mule/module/launcher/artifact/ArtifactFactoryUtils.java]: Agreed, but this is shared with the app parser. Are you saying it would make more sense to restrict it for domains?

File: modules/launcher/src/main/java/org/mule/module/launcher/descriptor/DomainDescriptor.java (hunk 12 above)
Comment (afelisatti) [inline modules/launcher/src/main/java/org/mule/module/launcher/descriptor/DomainDescriptor.java]: Not really, it just makes the hierarchy clearer and would later allow to have domain specific properties easily.

File: modules/launcher/src/test/java/org/mule/module/launcher/domain/DefaultDomainFactoryTestCase.java [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+        Domain domain = domainFactory.createArtifact(name);
+        assertThat(domain.getArtifactName(), is(name));
+        assertThat(domain.getDescriptor().getName(), is(name));
+        assertThat(domain.getDescriptor().isRedeploymentEnabled(), is(redeployment));
+        assertThat(domain.getDescriptor().getLoaderOverride(), loaderOverridesMatcher);
+    }
+
+    private void createDeployPropertiesFile(String domainName) throws FileNotFoundException, UnsupportedEncodingException
+    {
+        File properties = new File(muleHomeFolder + File.separator + MULE_DOMAIN_FOLDER + File.separator + domainName, DEFAULT_DEPLOY_PROPERTIES_RESOURCE);
```
Comment (pablokraan) [inline modules/launcher/src/test/java/org/mule/module/launcher/domain/DefaultDomainFactoryTestCase.java]: Use MuleFolderUtils for this

File: modules/launcher/src/test/java/org/mule/module/launcher/domain/DefaultDomainFactoryTestCase.java [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+        PrintWriter writer = new PrintWriter(properties, "UTF8");
+        writer.println("redeployment.enabled=false");
```
Comment (pablokraan) [inline modules/launcher/src/test/java/org/mule/module/launcher/domain/DefaultDomainFactoryTestCase.java]: Use the property constants

File: modules/launcher/src/main/java/org/mule/module/launcher/DeploymentDirectoryWatcher.java (hunk 3 above)
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/DeploymentDirectoryWatcher.java]: I know, but the original E/R was about "Allow domains to use loader.override"

File: modules/launcher/src/main/java/org/mule/module/launcher/artifact/Artifact.java [hunk 25]
```diff
@@ -9,13 +9,14 @@
...
 import org.mule.module.launcher.InstallException;
+import org.mule.module.launcher.descriptor.ArtifactDescriptor;
 
 import java.io.File;
 
 /**
  * An Artifact is an abstract representation of a deployable unit within the mule container.
  */
-public interface Artifact
+public interface Artifact<D extends ArtifactDescriptor>
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/artifact/Artifact.java]: Update the javadoc to include D

File: modules/launcher/src/main/java/org/mule/module/launcher/DeploymentDirectoryWatcher.java [hunk 26]
```diff
@@ -522,20 +522,26 @@ private void deleteAnchorsFromDirectory(final File directory)
...
+        redeployModifiedArtifacts(redeployableDomains, domainTimestampListener, domainArchiveDeployer);
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/DeploymentDirectoryWatcher.java]: Makes sense to be able to redeploy a domain's app but no the domain?

File: modules/launcher/src/main/java/org/mule/module/launcher/artifact/ArtifactFactoryUtils.java (hunk 9 above)
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/artifact/ArtifactFactoryUtils.java]: This is something I won't add for domain for sure. As this is for 3.8, I think it can also be removed for apps. Maybe you can check this with the rest of the team

File: modules/launcher/src/main/java/org/mule/module/launcher/descriptor/ArtifactDescriptor.java [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+ */
+package org.mule.module.launcher.descriptor;
+
+import org.mule.module.launcher.artifact.Artifact;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Represents an abstract description of an {@link Artifact}
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/descriptor/ArtifactDescriptor.java]: What does means "abstract" here? (asking as the class is not abstract)

Comment (afelisatti) [inline modules/launcher/src/main/java/org/mule/module/launcher/descriptor/ArtifactDescriptor.java]: Just that's the basis of the rest. Probably "base description" is more accurate.

File: modules/launcher/src/main/java/org/mule/module/launcher/descriptor/DomainDescriptorParser.java [hunk 28]
```diff
@@ -0,0 +1,48 @@
...
+ */
+public class DomainDescriptorParser
+{
+
+    public DomainDescriptor parse(File descriptor) throws IOException
+    {
+        final Properties properties = loadProperties(new FileInputStream(descriptor));
+        DomainDescriptor domainDescriptor = new DomainDescriptor();
+
+        domainDescriptor.setRedeploymentEnabled(BooleanUtils.toBoolean(properties.getProperty(PROPERTY_REDEPLOYMENT_ENABLED, Boolean.TRUE.toString())));
```
Comment (pablokraan) [inline modules/launcher/src/main/java/org/mule/module/launcher/descriptor/DomainDescriptorParser.java]: Use statics imports

File: modules/launcher/src/main/java/org/mule/module/launcher/DeploymentDirectoryWatcher.java (hunk 26 above)
Comment (afelisatti) [inline modules/launcher/src/main/java/org/mule/module/launcher/DeploymentDirectoryWatcher.java]: Yes, any combination should be valid. It would be up to the user to take advantage of this. By default all domains will redeploy though so the past behaviour is guaranteed.

File: modules/launcher/src/test/java/org/mule/module/launcher/builder/DomainFileBuilder.java (hunk 20 above)
Comment (afelisatti) [inline modules/launcher/src/test/java/org/mule/module/launcher/builder/DomainFileBuilder.java]: This was fixed in the latest commit.

File: modules/launcher/src/test/java/org/mule/module/launcher/DeploymentServiceTestCase.java [hunk 29]
```diff
@@ -2350,17 +2373,22 @@ private void changeConfigFile(String applicationPath, String configFile) throws
...
 
     private void doRedeployDummyDomainByChangingConfigFileWithBadOne() throws URISyntaxException, IOException
     {
-        doRedeployDummyDomainByChangingConfigFile("/bad-domain-config.xml");
+        doRedeployDomainByChangingConfigFile("/bad-domain-config.xml", dummyDomainFileBuilder);
     }
 
-    private void doRedeployDummyDomainByChangingConfigFile(String configFile) throws URISyntaxException, IOException
+    private void doRedeployDomainByChangingConfigFile(String configFile, DomainFileBuilder domain) throws URISyntaxException, IOException
     {
```
Comment (pablokraan) [inline modules/launcher/src/test/java/org/mule/module/launcher/DeploymentServiceTestCase.java]: There is no deployment test for the domain's loader override change
